### PR TITLE
Add another mask(upper tringle from bottom right) to flash attetion

### PR DESCRIPTION
**Pull request by ltqin**

### CODE REVIEW

**Comment by danyao12** [comment]
Maybe we should rename original MaskOutUpperTriangle to XXXFromTopLeft, and modify the use of mask in all related examples and device ops.

**Comment by ltqin** [comment]
> 

name to MaskUpperTriangleFromTopLeft?

**Comment by ltqin** [comment]
> Maybe we should rename original MaskOutUpperTriangle to XXXFromTopLeft, and modify the use of mask in all related examples and device ops.

name to MaskUpperTriangleFromTopLeft?

**Comment by danyao12** [comment]
> > Maybe we should rename original MaskOutUpperTriangle to XXXFromTopLeft, and modify the use of mask in all related examples and device ops.
> 
> name to MaskUpperTriangleFromTopLeft?

MaskUpperTriangleFromTopLeft/MaskUpperTringleFromBottomRight or just like xformers CausalFromTopLeft/CausalFromBottomRight.

File: include/ck/tensor_operation/gpu/device/masking_specialization.hpp [hunk 1]
```diff
@@ -47,13 +50,39 @@ struct MaskOutUpperTrianglePredicate
         return operator()(m + m_tile - 1, n);
     }
 };
+struct MaskUpperTriangleFromBottomRightPredicate
+{
+    MaskUpperTriangleFromBottomRightPredicate() : offset_(0) {}
+    __host__ __device__ void SetOffset(const index_t offset) { offset_ = offset; }
```
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/device/masking_specialization.hpp]: Rename `offset`  and `offset_` to `diagonal_offset` and `diagonal_offset_`  respectively,   and add some comments to show the user how to calculate the `diagonal_offset`

File: include/ck/tensor_operation/gpu/device/masking_specialization.hpp [hunk 2]
```diff
@@ -47,13 +50,39 @@ struct MaskOutUpperTrianglePredicate
...
-    C0MatrixMask_impl(index_t NRaw) : NRaw_(NRaw), predicate_(MaskOutPredicate{}) {}
+    C0MatrixMask_impl(index_t MRaw, index_t NRaw) : NRaw_(NRaw), predicate_(MaskOutPredicate{})
+    {
+        if constexpr(std::is_same<MaskOutPredicate,
+                                  MaskUpperTriangleFromBottomRightPredicate>::value)
+        {
+            if(NRaw > MRaw)
+                predicate_.SetOffset(NRaw - MRaw);
+        }
+    }
```
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/device/masking_specialization.hpp:85]: Suggest:  don't call `predicate_.SetOffset()` here, assume the `predicate` is already completely constructed,  just pass and store the predicate object here.  And for  `MaskUpperTriangleFromBottomRightPredicate`, remove `SetOffset`, the constructor is enough, cause it is the user to provide correct diagonal_offset

File: include/ck/tensor_operation/gpu/device/masking_specialization.hpp (hunk 1 above)
Comment (qianfengz) [reply include/ck/tensor_operation/gpu/device/masking_specialization.hpp]: Also rename `SetOffset` to `SetDiagonalOffset`

File: include/ck/tensor_operation/gpu/device/masking_specialization.hpp [hunk 3]
```diff
@@ -47,13 +50,39 @@ struct MaskOutUpperTrianglePredicate
...
+{
+    MaskUpperTriangleFromBottomRightPredicate() : offset_(0) {}
+    __host__ __device__ void SetOffset(const index_t offset) { offset_ = offset; }
+    __host__ __device__ constexpr bool operator()(index_t m, index_t n) const
+    {
+        return n > (m + offset_);
+    }
+
+    __host__ __device__ constexpr bool
+    IsTileSkippable(index_t m, index_t n, index_t m_tile, index_t /*n_tile*/) const
```
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/device/masking_specialization.hpp]: Rename `m` to `m_tile_orig`, `n` to `n_tile_orig`, `m_tile` to `m_tile_size`, `n_tile` to `n_tile_size`

File: include/ck/tensor_operation/gpu/device/masking_specialization.hpp (hunk 1 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/device/masking_specialization.hpp]: done

File: include/ck/tensor_operation/gpu/device/masking_specialization.hpp (hunk 3 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/device/masking_specialization.hpp]: done

**Review by qianfengz — APPROVED** [review]

**Review by danyao12 — APPROVED** [review]

**Pull request merged**